templates/ext/prism/api_node.c.erb

```diff
@@ -36,18 +36,26 @@ pm_string_new(pm_string_t *string, rb_encoding *encoding) {
     return rb_enc_str_new((const char *) pm_string_source(string), pm_string_length(string), encoding);
 }
 
-// Create a Prism::Source object from the given parser.
+// Create a Prism::Source object from the given parser, after pm_parse() was called.
 VALUE
 pm_source_new(pm_parser_t *parser, rb_encoding *encoding) {
-    VALUE source = rb_enc_str_new((const char *) parser->start, parser->end - parser->start, encoding);
-    VALUE offsets = rb_ary_new_capa(parser->newline_list.size);
+    VALUE source_string = rb_enc_str_new((const char *) parser->start, parser->end - parser->start, encoding);
+    VALUE source_argv[] = { source_string };
+    VALUE source = rb_class_new_instance(1, source_argv, rb_cPrismSource);
+
+    pm_source_init(source, parser);
+    return source;
+}
 
+void
+pm_source_init(VALUE source, pm_parser_t *parser) {
+    rb_funcall(source, rb_intern("start_line="), 1, LONG2NUM(parser->start_line));
```

Comment:
I'd prefer not to go through the ceremony of making two method calls in this function. We already have the start line when the source is created, and when we create the array for the offsets we can keep that reference around and push directly into it. That'll avoid the frame and everything else that comes from `rb_funcall`.

Reply:
> We already have the start line when the source is created,

This is handled here to be consistent with how Loader.java and FFI are doing things on master (and there it's not known before what is the start line, we know it after deserializing it).
https://github.com/ruby/prism/blob/18ada3120231e5b7faa821b0710f5c19f79fa11a/templates/lib/prism/serialize.rb.erb#L81-L83

I think it's valuable to be consistent here to avoid errors and help understanding how it all works.

> when we create the array for the offsets we can keep that reference around and push directly into it

If we do that then we cannot use `rb_ary_new_capa(parser->newline_list.size)` and so it will occur extra resizes (so extra allocations and memory copies).

> I'd prefer not to go through the ceremony of making two method calls in this function.

I can change them to `rb_iv_set()` if you want. But that would be inconsistent with how FFI does things.

Reply:
In that case you can call `rb_ary_resize`. I don't want us to be setting instance variables later if we don't have to.

Reply:
That's not available for the corresponding piece of code in `serialize.rb.erb`.

I want `@start_line` and `@offsets` to be handled consistently for all languages and all places setting it (for readability, maintainability and avoiding bugs), and for `@start_line` it needs to be assigned (e.g. for `Parser.load`), there is no alternative.
I think that is reasonable and more important than avoiding to assign ivars (for this case).

BTW the Java code deserializing the offsets naturally returns a new array, so it needs to assign, no alternative possible there.

I do agree setting instance variables later is not ideal in general, but there is no way to avoid mutating the Source object (thinking of it as an object graph, which includes the offsets) after it's built here, because the offsets are only available after parsing. I think this mutation is harmless since the user cannot access the Source object in an incomplete state.
I refactored the Java API a bit specifically to have the same property of no incomplete Source object accessible by the user.

Reply:
As an alternative, if you cannot have it be inconsistent, please make the Java side work like the CRuby side.

Reply:
Done.
To my surprise, it is indeed measurably faster for `Prism.parse("1 + 2")` (see commit message for the results).
I would have expected these 2 calls to rb_iv_set()/rb_funcall() to be insignificant (and they might be for real files with more core/many lines, but not for "1 + 2").
Also interesting is that `rb_iv_set()` is quite slow on CRuby, to the point that `rb_funcall()` (for an attr_writer) is twice as fast: https://gist.github.com/eregon/a6404a71151b30021ca081904f7149a0

OTOH for serialize.rb, the fastest would be public `attr_writer`.